File: packages/java/src/main/java/fyi/sererr/SourceContext.java
package fyi.sererr;

import java.util.List;
import java.util.Optional;

/**
 * Capture-time helper that populates a {@link StackFrame}'s
 * {@code contextLine} / {@code preContext} / {@code postContext}
 * from a {@link SourceProvider}.
 *
 * <p>{@link StackFrame} is immutable; this helper returns a new frame
 * with the populated fields. It is a no-op when:
 * <ul>
 *   <li>{@link StackFrame#line()} is 0 (unknown), OR</li>
 *   <li>the provider returns {@link Optional#empty()} for the file, OR</li>
 *   <li>the source is shorter than {@code frame.line()}.</li>
 * </ul>
 *
 * <p>{@code surrounding} is the number of lines before / after the
 * target line to capture; 5 is a sensible default (matches Sentry's UI).
 */
public final class SourceContext {
    private SourceContext() {}

    /**
     * Populate the source-context fields of {@code frame} using
     * {@code provider}; returns a new {@link StackFrame} with the
     * fields populated, or {@code frame} unchanged when population is
     * not possible.
     */
    public static StackFrame populateContext(StackFrame frame, SourceProvider provider, int surrounding) {
        if (frame.line() == 0) {
            return frame;
        }
        Optional<String> source = provider.getSource(frame.file());
        if (source.isEmpty()) {
            return frame;
        }
        // Split on \n only (proto3 semantics: bytes are bytes). Use a
        // -1 limit so trailing empty lines are preserved.
        String[] lines = source.get().split("\\R", -1);
        int idx = Math.max(0, frame.line() - 1);
        if (idx >= lines.length) {
            return frame;
        }
        String contextLine = lines[idx];
        int start = Math.max(0, idx - surrounding);
        int end = Math.min(lines.length, idx + 1 + surrounding);
        List<String> pre = List.of(java.util.Arrays.copyOfRange(lines, start, idx));
        List<String> post = List.of(java.util.Arrays.copyOfRange(lines, idx + 1, end));
        return StackFrame.builder()
                .function(frame.function())
                .module(frame.module())
                .pkg(frame.pkg())
                .file(frame.file())
                .absPath(frame.absPath())
                .line(frame.line())
                .contextLine(contextLine)
                .preContext(pre)
                .postContext(post)
                .sourceLink(frame.sourceLink())
                .inApp(frame.inApp())
                .build();
    }
}

File: packages/csharp/src/SourceContext.cs
namespace Sererr;

/// <summary>
/// Populates a <see cref="StackFrame"/>'s source-context fields from an
/// <see cref="ISourceProvider"/>. Records are immutable, so we return a
/// new frame rather than mutating in place.
/// </summary>
public static class SourceContext
{
    /// <summary>
    /// Returns a copy of <paramref name="frame"/> with
    /// <c>ContextLine</c>, <c>PreContext</c> and <c>PostContext</c>
    /// populated from <paramref name="provider"/>.
    ///
    /// No-op (returns <paramref name="frame"/> unchanged) when the
    /// provider has no source for the file, when the frame's line is 0,
    /// or when the line number exceeds the file's length.
    /// </summary>
    /// <param name="surrounding">
    /// Number of lines of pre/post context to capture (5 is a sensible
    /// default — matches Sentry's UI).
    /// </param>
    public static StackFrame PopulateContext(
        StackFrame frame,
        ISourceProvider provider,
        int surrounding)
    {
        if (frame.Line == 0)
        {
            return frame;
        }
        var source = provider.GetSource(frame.File);
        if (source is null)
        {
            return frame;
        }
        // Match Rust's `str::lines()` which strips trailing \n / \r\n
        // and drops a trailing empty line.
        var lines = SplitLines(source);
        var idx = checked((int)frame.Line) - 1;
        if (idx < 0 || idx >= lines.Count)
        {
            return frame;
        }
        var start = Math.Max(0, idx - surrounding);
        var endExclusive = Math.Min(lines.Count, idx + 1 + surrounding);
        return frame with
        {
            ContextLine = lines[idx],
            PreContext = lines.GetRange(start, idx - start),
            PostContext = lines.GetRange(idx + 1, endExclusive - (idx + 1)),
        };
    }

    private static List<string> SplitLines(string source)
    {
        // Rust's `str::lines()`:
        //   "a\nb\n".lines() -> ["a", "b"]
        //   "a\n".lines() -> ["a"]
        //   "".lines() -> []
        var result = new List<string>();
        var start = 0;
        for (var i = 0; i < source.Length; i++)
        {
            if (source[i] == '\n')
            {
                var lineEnd = i;
                if (lineEnd > start && source[lineEnd - 1] == '\r') lineEnd--;
                result.Add(source.Substring(start, lineEnd - start));
                start = i + 1;
            }
        }
        if (start < source.Length)
        {
            result.Add(source.Substring(start));
        }
        return result;
    }
}
